### PR TITLE
Multi-targeting with TargetFrameworks

### PR DESCRIPTION
**Pull request by michaelgsharp**
With .NET Core 2 no longer in support, we needed to fix the build configurations and frameworks.

This PR includes all the changes to switch from the way we used build configuration to specify the target framework for the tests and removes .NET Core 2.1. These changes include:
* No longer using the build configuration to specify the framework (Debug-netcoreapp3_1/Debug-netfx for their associated frameworks) to just using `<TargetFrameworks>` for the test projects. 
* Removing all build configurations beyond Debug/Release.
* Removing .NET Core 2.1 as part of this process, so currently only .NET Framework 4.6.1 and  .NET Core 3.1 are used. (.NET 6 will be added in a later PR as it changes some test results slightly).

### CODE REVIEW

**Comment by michaelgsharp** [comment]
The work to switch to multi targeting and not use the build configs is finished. This should actually raise our test coverage, as I found a bug during the process that was excluding a couple any of the tests that only were running on .netcore 3.1. Not sure how much of a bump it will end up being though.

@eerhardt Will this make getting AOT done easier? Or do we need to add the multi targeting to the actual src's and not just the tests?

File: .vsts-dotnet-ci.yml [hunk 1]
```diff
@@ -66,25 +66,14 @@ jobs:
     name: Centos_x64_NetCoreApp31
     buildScript: ./build.sh
     container: CentosContainer
```
Comment (safern) [inline .vsts-dotnet-ci.yml:68]: Why are we removing the Centos matrixes?

Comment (michaelgsharp) [reply .vsts-dotnet-ci.yml:68]: Those are now the default options so we don't need to specify them explicitly.

File: .vsts-dotnet-ci.yml [hunk 2]
```diff
@@ -177,11 +144,11 @@ jobs:
 
 - template: /build/ci/job-template.yml
   parameters:
-    name: Windows_x86_NetCoreApp21
+    name: Windows_x86_NetCoreApp31
     architecture: x86
     buildScript: build.cmd
     innerLoop: true
-    vsTestConfiguration: "/Framework:.NETCoreApp,Version=v2.1"
+    vsTestConfiguration: "/Framework:.NETCoreApp,Version=v3.0"
```
Comment (eerhardt) [inline .vsts-dotnet-ci.yml:151]: Why is this 3.0 and not 3.1?

File: docs/building/netcoreapp3.1-instructions.md [hunk 3]
```diff
@@ -1,14 +1,9 @@
-ML.NET source code files build for .NET Core 3.1 and .NET Standard 2.0. However, ML.NET test files only build on one target framework below at a time:
-- .NET Core 2.1
-- .NET Core 3.1
-- .NET Framework 4.6.1
+ML.NET source and test code files build for .NET Core 3.1 and NET Framework 4.6.1.
 
-To run tests on .NET Core 3.1, you need to do a few manual steps.
+If you want to run command line tests on only 1 of the frameworks or on a different framework, do this:
+ add `/p:TestTargetFramework=<Framework Names>` to the command line build.
```
Comment (eerhardt) [inline docs/building/netcoreapp3.1-instructions.md:4]: Maybe this doesn't need to happen on this first PR, but it would be best/nice if we supported the "normal" way of `dotnet test -f <FRAMEWORK>`

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <CopyLocalLockFileAssemblies>false</CopyLocalLockFileAssemblies>
```
Comment (eerhardt) [inline docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:6]: Why is CopyLocalLockFileAssemblies=false added to these samples? Do they still run?

File: src/Native/CMakeLists.txt [hunk 5]
```diff
@@ -149,11 +155,20 @@ endif(CMAKE_SYSTEM_NAME STREQUAL SunOS)
...
-            find_program(DSYMUTIL dsymutil)
-        else(CMAKE_SYSTEM_NAME STREQUAL Darwin)
-            find_program(OBJCOPY objcopy)
-        endif(CMAKE_SYSTEM_NAME STREQUAL Darwin)
+
+        if(CMAKE_SYSTEM_NAME STREQUAL Linux AND DEFINED CMAKE_TOOLCHAIN_FILE)
+            find_program(OBJCOPY
+                NAMES llvm-objcopy
+                llvm-objcopy${CLR_CMAKE_COMPILER_FILE_NAME_VERSION})
+            #find_program(OBJCOPY objcopy HINTS ${CROSS_ROOTFS}/usr/bin)
```
Comment (eerhardt) [inline src/Native/CMakeLists.txt:163]: Why is this commented out?

File: test/Directory.Build.props [hunk 6]
```diff
@@ -2,13 +2,14 @@
...
-    <TargetFramework Condition="$(Configuration.EndsWith('-netcoreapp3_1'))">netcoreapp3.1</TargetFramework>
-    <TargetFramework Condition="'$(TestTargetFramework)' != ''">$(TestTargetFramework)</TargetFramework>
+    <TargetFrameworks>netcoreapp3.1;net461</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp3.1</TargetFrameworks>
+    <TargetFrameworks Condition="$([MSBuild]::IsOSPlatform('OSX')) AND '$(TargetArchitecture)' == 'arm64'">net6.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(TestTargetFramework)' != ''">$(TestTargetFramework)</TargetFrameworks>
     <RuntimeIdentifier Condition="'$(TargetFramework)' == 'net461'">win-x64</RuntimeIdentifier>
     <IsPackable>false</IsPackable>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PreserveCompilationContext>true</PreserveCompilationContext>
```
Comment (eerhardt) [inline test/Directory.Build.props:12]: Why is this change needed?

File: test/Directory.Build.props [hunk 7]
```diff
@@ -2,13 +2,14 @@
   <Import Project="..\Directory.Build.props" />
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
-    <TargetFramework Condition="$(Configuration.EndsWith('-netfx'))">net461</TargetFramework>
-    <TargetFramework Condition="$(Configuration.EndsWith('-netcoreapp3_1'))">netcoreapp3.1</TargetFramework>
-    <TargetFramework Condition="'$(TestTargetFramework)' != ''">$(TestTargetFramework)</TargetFramework>
+    <TargetFrameworks>netcoreapp3.1;net461</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp3.1</TargetFrameworks>
+    <TargetFrameworks Condition="$([MSBuild]::IsOSPlatform('OSX')) AND '$(TargetArchitecture)' == 'arm64'">net6.0</TargetFrameworks>
```
Comment (eerhardt) [inline test/Directory.Build.props:7]: Are we building for OSX-arm64? I didn't think we were.

File: .vsts-dotnet-ci.yml (hunk 2 above)
Comment (michaelgsharp) [reply .vsts-dotnet-ci.yml:151]: I copied it from [here](https://github.com/dotnet/machinelearning/blob/793872853c4af040026a7825381be567ec25da92/.vsts-dotnet-ci.yml#L119) which was there already. Should it be 3.1? If so, I will go ahead and update both of them.

File: Directory.Build.props [hunk 8]
```diff
@@ -22,7 +22,7 @@
 
   <PropertyGroup>
     <Configuration Condition="'$(Configuration)'==''">Debug</Configuration>
-    <Configurations>Debug;Release;Debug-netcoreapp3_1;Release-netcoreapp3_1;Debug-netfx;Release-netfx</Configurations>
+    <Configurations>Debug;Release;</Configurations>
```
[the remark this answers is not in the document]
Comment (michaelgsharp) [reply Directory.Build.props:25]: Sounds good. Removed, code not pushed yet.

File: .vsts-dotnet-ci.yml (hunk 2 above)
Comment (eerhardt) [reply .vsts-dotnet-ci.yml:151]: I just found it odd that we use 3.1 everywhere else, but 3.0 here...

File: docs/building/netcoreapp3.1-instructions.md (hunk 3 above)
Comment (michaelgsharp) [reply docs/building/netcoreapp3.1-instructions.md:4]: This does work on individual projects just fine. These instructions are for the `build.cmd` script itself. Is there a better way to handle this with a build script and not just directly with `dotnet test`?

File: src/Native/CMakeLists.txt (hunk 5 above)
Comment (michaelgsharp) [reply src/Native/CMakeLists.txt:163]: Good catch. That was from my testing and is now removed, code not pushed yet.

File: test/Directory.Build.props (hunk 7 above)
Comment (michaelgsharp) [reply test/Directory.Build.props:7]: We are. The M1 changes went in with the 1.6.0 release.

File: docs/building/netcoreapp3.1-instructions.md (hunk 3 above)
Comment (eerhardt) [reply docs/building/netcoreapp3.1-instructions.md:4]: I almost wonder if this page is still needed going forward... thoughts on removing it?

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj (hunk 4 above)
Comment (michaelgsharp) [reply docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:6]: Yeah they still run. Before when they were .netcore2 they didn't copy the nuget runtime folder into the output directory. I'm guessing the default changed with .netcore3 as the runtime folders are now copied into the output directory. This was causing issues during the CI process as it increased the space needed by like 3GB and that put us over the limit on the machines. This is preventing that copy from happening.

File: test/Directory.Build.props (hunk 6 above)
Comment (michaelgsharp) [reply test/Directory.Build.props:12]: Without it, the `<Assembly name>.deps.json` isn't created. Currently the helix testing needs that file.

File: Directory.Build.props (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelgsharp) [reply Directory.Build.props]: Code pushed.

File: src/Native/CMakeLists.txt (hunk 5 above)
Comment (michaelgsharp) [reply src/Native/CMakeLists.txt]: Code pushed.

File: test/Directory.Build.props (hunk 6 above)
Comment (eerhardt) [reply test/Directory.Build.props:12]: This is the wrong property to get a `.deps.json` file created. It makes too much stuff.

Instead we should use `GenerateDependencyFile=true`.

File: docs/building/netcoreapp3.1-instructions.md (hunk 3 above)
Comment (michaelgsharp) [reply docs/building/netcoreapp3.1-instructions.md:4]: I think we are good to remove it. since 3.1 is now default we don't need separate instructions.

File: test/Directory.Build.props (hunk 6 above)
Comment (michaelgsharp) [reply test/Directory.Build.props:12]: Good to know. Switched it.

File: .vsts-dotnet-ci.yml (hunk 2 above)
Comment (michaelgsharp) [reply .vsts-dotnet-ci.yml:151]: I have switched it to 3.1.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!
